gammapy/utils/fitting/sampling.py

```diff
@@ -0,0 +1,162 @@
+
+
+# Read/write parameters in the dataset
+def model_to_par(dataset):
+    """
+    Return a tuple of the factor parameters of all
+    free parameters in the dataset sky model.
+    """
+    pars = []
+    for p in dataset.parameters.free_parameters:
+        pars.append(p.factor)
+
+    return pars
+
+
+def par_to_model(dataset, pars):
+    """Update model in dataset with a list of free parameters factors"""
+    for i, p in enumerate(dataset.parameters.free_parameters):
+        p.factor = pars[i]
+
+
+# Compute LogLike associated with prior and data/model evaluation
+def lnprior(dataset):
+    """
+    Return probability of parameter values according to prior knowledge.
+    Parameter limits should be done here through uniform prior ditributions
+    """
+    logprob = 0
+    for par in dataset.parameters.free_parameters:
+        logprob += uniform_prior(par.value, par.min, par.max)
+
+    return logprob
+
+
+def lnprob(pars, dataset):
+    """Estimate the likelihood of a model including prior on parameters."""
+    # Update model parameters factors inplace
+    for factor, par in zip(pars, dataset.parameters.free_parameters):
+        par.factor = factor
+
+    lnprob_priors = lnprior(dataset)
+
+    # dataset.likelihood returns Cash statistics values
+    # emcee will maximisise the LogLikelihood so we need -dataset.likelihood
+    total_lnprob = -dataset.likelihood() + lnprob_priors
+
+    return total_lnprob
+
+
+def run_mcmc(dataset, nwalkers=8, nrun=1000, threads=1):
+    """
+    Run the MCMC sampler.
+
+    Parameters
+    ----------
+    dataset : `MapDataset`
+        A gammapy dataset object. This contains the observed counts cube,
+        the exposure cube, the psf cube, and the sky model and model.
+        Each free parameter in the sky model is considered as parameter for the MCMC.
+    nwalkers: int
+        Required integer number of walkers to use in ensemble.
+        Minimum is 2*nparam+2, but more than that is usually better.
+        Must be even to use MPI mode.
+    nrun: int
+        Number of steps for walkers. Typically at least a few hundreds (but depends on dimensionality).
+        Low nrun (<100?) will underestimate the errors.
+        Samples that would populate the distribution are nrun*nwalkers.
+        This step can be ~seen as the error estimation step.
+    """
+    dataset.parameters.autoscale()  # Autoscale parameters
+    pars = model_to_par(dataset)  # get a tuple of parameters from dataset
+    ndim = len(pars)
+
+    # Initialize walkers in a ball of relative size 0.5% in all dimensions if the
+    # parameters have been fit, or to 10% otherwise
+    spread = 0.5 / 100
+    p0var = np.array([spread * pp for pp in pars])
+    p0 = emcee.utils.sample_ball(pars, p0var, nwalkers)
+
+    labels = []
+    for par in dataset.parameters.free_parameters:
+        labels.append(par.name)
+        if (par.min is np.nan) and (par.max is np.nan):
+            print(
```

Comment:
Always use logging like elsewhere in Gammapy. Never print from a library. This gives the library user or others building packages on top of a library control.

Reply:
Done.
Is there a way to activate the showing of logging warning and info in the notebooks so that the progression can be shown in the notebook ?

Reply:
Putting this at the top of your script and notebook should work:
```
import logging
logging.basicConfig(level=logging.DEBUG)
```
Only the first `basicConfig` call wins within a given Python interpreter run, so this should NEVER appear in libraries, and it should be at the top of your scripts.